### PR TITLE
array: fix `asarray` for array input with `dtype`

### LINKED ISSUE #11285: BUG: `array.asarray` does not respect `dtype` arg
**Describe the issue**:

`dask.array.asarray` does not respect the `dtype` argument.

**Minimal Complete Verifiable Example**:

```python
>>> import numpy as np
>>> import dask.array as da
>>> Zm = da.asarray([[1, 2, 3]]) 
>>> Zm
dask.array<array, shape=(1, 3), dtype=int64, chunksize=(1, 3), chunktype=numpy.ndarray>
>>> Z = da.asarray(Zm, dtype=da.float64)
>>> Z
dask.array<array, shape=(1, 3), dtype=int64, chunksize=(1, 3), chunktype=numpy.ndarray>
>>> Z.compute().dtype
dtype('int64')
# same issue is present with `np` dtypes directly
>>> Z = da.asarray(Zm, dtype=np.float64)
>>> Z
dask.array<array, shape=(1, 3), dtype=int64, chunksize=(1, 3), chunktype=numpy.ndarray>
>>> Z.compute().dtype
dtype('int64')
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2024.8.0+3.g65270980
- Python version: 3.12.4
- Operating System: Ubuntu
- Install method (conda, pip, source): `python -m pip install git+https://github.com/dask/dask.git`

### PR DESCRIPTION
**Pull request by lucascolley**
- [x] Closes #11285
- [x] Tests added / passed (haven't checked tests locally)
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

File: dask/array/core.py [hunk 1]
```diff
@@ -4575,12 +4575,12 @@ def asarray(
...
+        if isinstance(a, Array) and dtype is None and order is None:
             return a
-        elif hasattr(a, "to_dask_array"):
+        elif hasattr(a, "to_dask_array") and dtype is None and order is None:
             return a.to_dask_array()
         elif type(a).__module__.split(".")[0] == "xarray" and hasattr(a, "data"):
-            return asarray(a.data)
+            return asarray(a.data, dtype=dtype, order=order)
         elif isinstance(a, (list, tuple)) and any(isinstance(i, Array) for i in a):
             return stack(a, allow_unknown_chunksizes=allow_unknown_chunksizes)
```
Comment (lucascolley) [inline dask/array/core.py:4585]: we may want `dtype=dtype, order=order` here too

**Comment by lucascolley** [comment]
cc @jsignell @jakirkham @jrbourbeau from gh-8106

**Comment by lucascolley** [comment]
Hmm, it looks like `from_array` does not accept Dask arrays, redirecting to `asarray`, but in `asarray` `dtype` and `order` are ignored. I suppose we want to coerce the dtype and order from `asarray` then.

Should I try using `Array.astype`? And perhaps we don't need to allow changing order?

**Comment by lucascolley** [comment]
I've hopefully fixed the case of `Array` input with `dtype`. Otherwise, I have added the warning:

```
.. warning::
        `dtype` is ignored if `a` has the attribute ``to_dask_array`` or
        is a list or tuple of `Array`'s.
        `order` is ignored if `a` is an `Array`, has the attribute ``to_dask_array``,
        or is a list or tuple of `Array`'s.
```

**Comment by lucascolley** [comment]
Currently the warning is below Examples on the docs pages. Not sure if it is possible to move upwards while keeping the admonition.

File: dask/array/core.py [hunk 2]
```diff
@@ -4573,14 +4573,20 @@ def asarray(
...
+        is a list or tuple of `Array`'s.
+        `order` is ignored if `a` is an `Array`, has the attribute ``to_dask_array``,
+        or is a list or tuple of `Array`'s.
     """
     if like is None:
         if isinstance(a, Array):
-            return a
+            return a if dtype is None else a.astype(dtype)
         elif hasattr(a, "to_dask_array"):
             return a.to_dask_array()
```
Comment (lucascolley) [inline dask/array/core.py:4587]: I suppose we can use `astype` in this case too.

File: dask/array/core.py [hunk 3]
```diff
@@ -4641,14 +4647,20 @@ def asanyarray(a, dtype=None, order=None, *, like=None, inline_array=False):
...
+    .. warning::
+        `dtype` is ignored if `a` has the attribute ``to_dask_array`` or
+        is a list or tuple of `Array`'s.
+        `order` is ignored if `a` is an `Array`, has the attribute ``to_dask_array``,
+        or is a list or tuple of `Array`'s.
     """
     if like is None:
         if isinstance(a, Array):
-            return a
+            return a if dtype is None else a.astype(dtype)
```
Comment (phofl) [inline dask/array/core.py:4659]: Can you expand this into multiple lines? This breaks the pattern around it currently

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx

**Pull request merged**